### ISSUE TITLE
Guidance on Zipformer configuration for low-power ARM CPU (Persian ASR, ~400h data)

### ISSUE THREAD

**Issue by mohammadd-dev** [k2-fsa/icefall#2066]
Hi icefall team,

First of all, thank you for the great work on icefall and especially the Zipformer recipes. The modular design and Lhotse integration make it incredibly practical for adapting to new languages like Persian.

I am currently training a Persian (fa) ASR model using `egs/commonvoice/ASR/zipformer` as the base recipe, and I would really appreciate your insight on a few architectural and hyperparameter decisions.

---

### Goal

My target is **on-device inference** on a **low-power ARM CPU** with limited memory bandwidth.

**Priorities:**

* Small model footprint
* Low latency
* Robustness to noisy home environments

---

### Data & Preparation

* ~400 hours Persian speech (CommonVoice)
* Speed perturbation (0.9, 1.1) → ~1200h effective audio
* MUSAN noise/reverb augmentation enabled
* BPE size: 500

---

### Current “Small” Configuration (~18M parameters)

```
--encoder-dim "192, 256, 384, 256, 192"
--num-encoder-layers "2, 2, 3, 2, 2"
--feedforward-dim "512, 768, 1024, 768, 512"
--nhead "4, 4, 4, 4, 4"
--chunk-size 16
--left-context 64
```

---

### Questions

**1) Architecture choice**

For a small ARM CPU target, is Zipformer’s hierarchical structure expected to be significantly more cache-friendly during ONNX/NCNN inference compared to a flatter `stateless7` (Conv/Emformer-style) model?

---

**2) Downsampling strategy**

Would you recommend more aggressive `downsampling_factor` early in the network (e.g., 2 or 4) to reduce the sequence length seen by attention layers on CPU, or is it better to keep early layers at higher resolution for phonetic fidelity?

---

**3) BPE size**

Is BPE 500 too large for this model size?
Would reducing it to ~250–300 noticeably improve Joiner/Decoder speed for CPU inference?

---

**4) Latency vs. parameter count**

If I scale the model toward ~30–40M parameters (larger `encoder-dim`), is there a potential “sweet spot” where the model becomes more confident and reduces decoding/beam overhead enough to offset the larger matrix multiplications?

---

**5) Streaming vs. non-streaming**

For this type of hardware, would you recommend:

* Strict streaming (causal) Zipformer, or
* Non-streaming Zipformer with chunked attention as a compromise?

---

**6) Decoder choice**

Since this CPU has relatively slow memory access, would a simpler stateless decoder be noticeably faster than the default RNN-T decoder in this scenario?

---

I would also greatly appreciate any suggestions on how to better scale the Zipformer stacks or layer distribution to better fit the small cache and memory bandwidth of such CPUs.

Thank you again for this excellent toolkit.

Best regards

**Comment by csukuangfj** [comment]
https://k2-fsa.github.io/sherpa/ncnn/pretrained_models/zipformer-transucer-models.html#marcoyang-sherpa-ncnn-streaming-zipformer-20m-2023-02-17-english

Can you try this one?

and this one https://k2-fsa.github.io/sherpa/onnx/pretrained_models/online-transducer/zipformer-transducer-models.html#csukuangfj-sherpa-onnx-streaming-zipformer-en-20m-2023-02-17-english 
on your board and compare the RTF?

**Comment by mohammadd-dev** [comment]
**marcoyang/sherpa-ncnn-streaming-zipformer-20M-2023-02-17 (English)**

_Decode a single wave file:_

```
RTF ≈ 0.63 
wav duration (s): 11.1 s
Elapsed seconds: 6.987 s
Real time factor (RTF): 6.987 / 11.1 = 0.629
```

**Comment by csukuangfj** [comment]
Since the 20M model has a RTF < 1 on your board. Please follow how we train the 20M model.
